modo/cram.py

```diff
@@ -5,86 +5,117 @@
     AlignedSegment,
     AlignmentFile,
+    VariantFile,
+    VariantRecord,
 )
 from urllib.parse import urlparse
 import modo_schema.datamodel as model
 
 import sys
+import re
 import htsget
-from .helpers import parse_region, bytesio_to_alignment_segments
-from io import BytesIO
+from .helpers import (
+    parse_region,
+    get_fileformat,
+    make_pysam_inFile,
+    bytesio_to_iterator,
+    iter_to_file,
+)
+from io import BytesIO, StringIO
 
 
-def slice_cram(
+def slice_genomics(
     path: str,
-    region: Optional[str],
+    region: Optional[str] = None,
     reference_filename: Optional[str] = None,
     output_filename: Optional[str] = None,
-) -> Iterator[AlignedSegment]:
-    """Return an iterable slice of the CRAM file."""
+) -> Optional[Iterator[AlignedSegment | VariantRecord]]:
+    """Returns an iterable slice of the CRAM, VCF or BCF file,
+    or saves it to a local file."""
     if region:
-        chrom, start, end = parse_region(region)
+        reference_name, start, end = parse_region(region)
     else:
-        chrom, start, end = None, None, None
+        reference_name, start, end = None, None, None
+
+    fileformat = get_fileformat(path)
 
-    cramfile = AlignmentFile(path, "rc", reference_filename=reference_filename)
-    cram_iter = cramfile.fetch(chrom, start, end)
+    infile = make_pysam_inFile(
+        path=path, fileformat=fileformat, reference_filename=reference_filename
+    )
+
+    gen_iter = infile.fetch(reference_name, start, end)
 
     if output_filename:
-        output = AlignmentFile(
-            output_filename,
-            "wc",
-            template=cramfile,
+        iter_to_file(
+            gen_iter=gen_iter,
+            infile=infile,
+            output_filename=output_filename,
             reference_filename=reference_filename,
         )
-        for read in cram_iter:
-            output.write(read)
-        output.close()
 
-    return cram_iter
+    return gen_iter
 
 
-def slice_remote_cram(
+def slice_remote_genomics(
     url: str,
     region: Optional[str] = None,
     reference_filename: Optional[str] = None,
     output_filename: Optional[str] = None,
 ):
-    """Stream or write to a local file a slice of a remote CRAM file"""
+    """Stream or write to a local file a slice of a remote CRAM or VCF/BCF file"""
 
     url = urlparse(url)
+    in_fileformat = get_fileformat(url.path)
+    if in_fileformat not in ("CRAM", "VCF", "BCF") or url.path.endswith(
+        ".vcf"
+    ):
+        raise ValueError(
+            "Unsupported file type. Streaming/Saving remote genomic files support remote .cram, .vcf.gz, .bcf files."
+        )
+    # path = url.path
+    url = url._replace(
+        path=re.sub("\.vcf\.\w+$", ".vcf", url.path)
+    )  # remove aditional
+    # extensions, e.g., .vcf.gz -> .vcf
     url = url._replace(path=str(Path(url.path).with_suffix("")))
 
     if region:
         reference_name, start, end = parse_region(region)
     else:
```

Comment:
__Suggestion__: As this ifelse statement occurs multiple times (see `slice_genomics()`) it could be moved into `parse_region()`.

Reply:
Good point. 

Done - see commit https://github.com/sdsc-ordes/modo-api/commit/0702e34383980e4f31b701f7f73b6052dca36d3b.